### ISSUE TITLE
Support Numbering in Code Listings

### ISSUE THREAD

**Issue by devunwired** [asciidoctor/asciidoctor#1441]
Apologies if this is already discussed elsewhere, but I didn't see it in an issue search…

I would like to see the option for code listings to be auto-numbered the same way as figures and tables, so that these numbers show up in the listing titles and cross-references. This seems to be because the DocBook output places the xref id to a previous element instead of the `programlisting`:

```
<formalpara xml:id="my_xref_id">
<title>Code Listing Title</title>
<para>
<programlisting language="xml" linenumbering="unnumbered">
```

…as opposed to something like a table reference where the id is on the same element:

```
<table xml:id="table_versions" frame="all" rowsep="1" colsep="1">
```

**Comment by mojavelinux** [comment]
This issue is specific to the DocBook toolchain and is explained well on the following page: http://www.sagehill.net/docbookxsl/FormalTitles.html. It's not the placement of the id attribute but rather the choice of element (`formalpara` instead of `figure`).

There are two ways to approach this.

First, you can customize the template for the listing block (which is also the template for the source block) to output a different element (or even arrangement of elements). You can find a starting point here:

https://github.com/asciidoctor/asciidoctor-backends/blob/master/slim/docbook45/block_listing.xml.slim

Just copy that into a templates folder, tweak it and enable it using the `-T` option when calling Asciidoctor. See http://asciidoctor.org/docs/user-manual/#storing-multiple-templates (clearly not the best docs, but it should get you started).

Second, we could consider using `<figure>` element instead of `<formalpara>` when a certain attribute is set on the document (perhaps `listing-caption` to align with the HTML backend). The consequence, however, is that it will change the numbering of block images, throwing them all into one bucket. That may or may not be the desired result.

**Comment by mojavelinux** [comment]
Of course, a third option is to tweak the DocBook XSL, but I hesitate to recommend that path because it's a rather tedious ordeal. But it is an option.

**Comment by devunwired** [comment]
Thanks for the references, I'll look into those.

I do think that numbered listings ought to be separate from other elements (e.g. figures)…at least for my purposes. Based on the link you provided, I would almost think of a code listing as an `example` or `procedure`, rather than a `figure`. Are those elements already used elsewhere in the Asciidoctor toolchain?

**Comment by mojavelinux** [comment]
The `<example>` is already used. The `<procedure>` is available, but I'm not convinced it fits here.

Keep in mind you can wrap your source listing in an example block and then you'll get the numbering for free (at the cost of some additional framing around the element).

DocBook sort of backs us into a corner here if we want to use what is available out of the box. Of course, it's entirely possible to hack the DocBook XSL so that it recognizes an `<example>` or `<figure>` with a special role for listings. But that's not my department :)

**Comment by devunwired** [comment]
Cool, thanks. I kind of like wrapping the listings that should be numbered inside an example. It sets them apart visually and allows the flexibility for shorter code snippets (just a few lines) to appear without numbers. I'll probably go with that for now.

I might try to tweak the current visual style (I can do that with a template?) so the double framing isn't quite so pronounced, but I do like the fact that it is visually different from a plain source block.

**Comment by mojavelinux** [comment]
:+1: 

>  (I can do that with a template?)

It depends on what output we are talking about. For the HTML output (that's generated by Asciidoctor), you'd tweak the stylesheet. If it's after you generate the DocBook, that's all part of the DocBook toolchain. It's certainly possible. You'd need this reference: http://www.sagehill.net/docbookxsl/
